src/aten/sycl/DistanceKernel.cpp

```diff
@@ -0,0 +1,463 @@
+#pragma clang diagnostic push
+#pragma GCC diagnostic push
+// Avoid SYCL compiler return-type error
+#pragma clang diagnostic ignored "-Wreturn-type"
+#pragma GCC diagnostic ignored "-Wreturn-type"
+
+#include <ATen/ATen.h>
+#include <ATen/AccumulateType.h>
+#include "ATen/Dispatch.h"
+#include <ATen/ExpandUtils.h>
+#include <comm/Runtime.h>
+#include <comm/SYCLContext.h>
+#include <comm/SYCLHelpers.h>
+#include <comm/TensorInfo.h>
+
+
+namespace at::native::xpu {
+template <typename scalar_t>
+static double device_sqrt(scalar_t val) {
+  return std::sqrt(val);
+};
+
+template <typename scalar_t>
+class dists {
+ public:
+  static scalar_t sign(scalar_t val) {
+    return (0 < val) - (val < 0);
+  }
+};
+
+// Zero norm
+template <typename scalar_t>
+struct dists_zero {
+  static void inc(scalar_t& agg, const scalar_t diff, const scalar_t p) {
+    agg += diff != 0.0f;
+  }
+  static scalar_t finish(const scalar_t agg, const scalar_t p) {
+    return agg;
+  }
+  static void agg(scalar_t& update, const scalar_t other) {
+    update += other;
+  }
+};
+
+// One norm
+template <typename scalar_t>
+struct dists_one {
+  static void inc(scalar_t& agg, const scalar_t diff, const scalar_t p) {
+    agg += diff;
+  }
+  static scalar_t finish(const scalar_t agg, const scalar_t p) {
+    return agg;
+  }
+  static void agg(scalar_t& update, const scalar_t other) {
+    update += other;
+  }
+  static scalar_t backward(
+      const scalar_t diff,
+      const scalar_t grad,
+      const scalar_t dist,
+      const scalar_t p) {
+    return grad * dists<scalar_t>::sign(diff);
+  }
+};
+
+// Two norm
+template <typename scalar_t>
+struct dists_two {
+  static void inc(scalar_t& agg, const scalar_t diff, const scalar_t p) {
+    agg += diff * diff;
+  }
+  static scalar_t finish(const scalar_t agg, const scalar_t p) {
+    return device_sqrt<scalar_t>(agg);
+  }
+  static void agg(scalar_t& update, const scalar_t other) {
+    update += other;
+  }
+  static scalar_t backward(
+      const scalar_t diff,
+      const scalar_t grad,
+      const scalar_t dist,
+      const scalar_t p) {
+    return dist == 0.0f ? static_cast<scalar_t>(0) : grad * diff / dist;
+  }
+};
+
+// General p norm
+template <typename scalar_t>
+struct dists_p {
+  static void inc(scalar_t& agg, const scalar_t diff, const scalar_t p) {
+    // TODO:
+    // Here had an unknown bug which affected other code segments
+    // unexpectedly. See Jira:
+    // https://jira.devtools.intel.com/browse/PYTORCHDGQ-958 for details Below
```

Comment:
Which bug are you talking about? The link you provide is intel internal link and is not helpful to non-intel people who might review/try out this code. Can you, please, create github issue instead with the description?